Kconfig.nrf

```diff
@@ -61,7 +61,7 @@ config ENTROPY_GENERATOR
 	default y if BUILD_WITH_TFM || (SPM && SPM_SECURE_SERVICES)
 
 config PSA_WANT_ALG_CTR_DRBG
-	default y if ENTROPY_GENERATOR
+	default y if ENTROPY_GENERATOR && !TFM_MINIMAL
```

Comment:
This is bad practice.

Features (DRBG) should not know of their users (TFM_MINIMAL).

Users should be enabling the features they need (which is evidently not DRBG in this case).

Can we add the removal of this patch to the DoD of https://projecttools.nordicsemi.no/jira/browse/NCSDK-13595 @frkv  ?

Reply:
This is default y for all things besides TF-M minimal configuration. This is the defacto PRNG both in CSPRNG in Zephyr and for The CryptoCell driver as well as the mbed TLS project, so no.

Reply:
The reason for this being needed is that `Kconfig.nrf` is sourced after `Kconfig.mbedtls_minimal.defconfig`.

Reply:
@joerchan quite the opposite:
```h
menu "nrf (/projects/github/ncs/nrf)"
osource "/projects/github/ncs/nrf/Kconfig.nrf"
config ZEPHYR_NRF_MODULE
        bool
        default y
endmenu
menu "mcuboot (/projects/github/ncs/bootloader/mcuboot)"
osource "$(ZEPHYR_MCUBOOT_KCONFIG)"
config ZEPHYR_MCUBOOT_MODULE
        bool
        default y
endmenu
menu "mbedtls (/projects/github/ncs/mbedtls)"
osource "$(ZEPHYR_MBEDTLS_KCONFIG)"
config ZEPHYR_MBEDTLS_MODULE
        bool
        default y
endmenu
menu "trusted-firmware-m (/projects/github/ncs/modules/tee/tf-m/trusted-firmware-m)"
osource "$(ZEPHYR_TRUSTED_FIRMWARE_M_KCONFIG)"
config ZEPHYR_TRUSTED_FIRMWARE_M_MODULE
        bool
        default y
endmenu
....
```
https://github.com/nrfconnect/sdk-nrf/blob/2a2fe9c009f5dfda0e38c95afb67e92833f536cf/modules/modules.cmake#L8
https://github.com/nrfconnect/sdk-nrf/blob/2a2fe9c009f5dfda0e38c95afb67e92833f536cf/modules/trusted-firmware-m/Kconfig#L55

Reply:
My mistake, then I don't have an explanation for why it is needed. The mbedtls_minimal_defconfig default is not set.

Maybe it is because `modules/mbedtls/Kconfig.psa` is sourced later, but that sets no default y.
```
At modules/mbedtls/Kconfig.psa:8
Included via /home/joakim/repos/ncs/zephyr/Kconfig:8 -> Kconfig.zephyr:33 -> modules/Kconfig:80 -> modules/mbedtls/Kconfig:176
```

Reply:
> My mistake, then I don't have an explanation for why it is needed.

Probably because the default that is sourced **first** wins.
This is also why a board default can overrule any other default.

Reply:
I'll leave it as is, since I don't know how to fix this in a good practice way.

Reply:
I wonder, what's the reason for having this default in `Kconfig.nrf` in the first place, and not in nrfxlib, for example one of those two files:
https://github.com/nrfconnect/sdk-nrfxlib/blob/main/nrf_security/Kconfig
https://github.com/nrfconnect/sdk-nrfxlib/blob/main/nrf_security/Kconfig.psa

Reply:
That is a good question. I'm also wondering why this default is there to begin with, in addition to the one above about ENTROPY_GENERATOR.
I'm  suspecting these two can actually be removed, since nrf_security has a select on ENTROPY_GENERATOR, and the PSA_WANT_ALG_CTR_DBRG is being select'ed from other places.
Possibly ENTROPY_CC3XX should have a select PSA_WANT_ALG_CTR_DRBG instead?

@frkv or @Vge0rge Any input on this?

Reply:
I will have to test if this will work actually. I think one issue that we are trying to solve here is to set the correct defaults. The PSA_WANT configuration is placed in the mbedtls repo from zephyr. The entropy generator one is sourced from zephyr. So either we need downstream patches if we need to solve this or we need to make sure that the nrfxlib Kconfig is sourced first, which I would assume that is not possible.

Reply:
Combined knowledge about TF-M and Mbed TLS is somewhat spread around in multiple domains at the moment, so the configurations can be placed in multiple locations. What we do with the override is **very explicit**, so I argue that it isn't an anti-pattern. It is using default-overrides as they are expected to be used.

PSA_WANT_ALG_CTR_DRBG and PSA_WANT_ALG_HMAC_DRBG is an invention made by us, and it covers a feature that is not finalized inside the PSA Crypto specifications. There exists no driver-support for CTR_DRBG/HMAC_DRBG as well as no configuration to state with clear intent that you wish to provide (and use) one or the other. We want to upstream these config-additions (once PSA driver support is in place).

CTR_DRBG has a much higher impact (in code) than HMAC_DRBG. But more importantly CTR_DRBG is also the default solution for PRNG in our CryptoCell deliverables (and in fact inside regular builds of Mbed TLS as well)

Having HMAC_DRBG as the "only option" is a use-case limited to TF-M minimal configuration (hence the override of the default).

ENTROPY_CC3XX should not select PSA_WANT_ALG_CTR_DRBG. The presence of ENTROPY_CC3XX is only a statement "I want to provide entropy".  The choice of PRNG isn't made by enabling/disabling this.

Reply:
Since we do enable the CTR_DRBG in nrfxlib by default: 
https://github.com/nrfconnect/sdk-nrfxlib/blob/c72e91d8014fe257fa4112664d213d80ae30c8c6/nrf_security/Kconfig#L189

And we also enable entropy when nrf_security is enabled: 
https://github.com/nrfconnect/sdk-nrfxlib/blob/c72e91d8014fe257fa4112664d213d80ae30c8c6/nrf_security/Kconfig#L37

Then we do we explicitly need these overrides though?  Which corner case we are trying to solve? 
I removed these from Kconfig.nrf and built some crypto samples in secure/non-secure targets and these Kconfig were always enabled.

Reply:
Ah. The MBEDTLS_CTR_DRBG_ENABLED is a copy of configurations from Zephyr. This should not be default y

This isn't even a valid Mbed TLS configuration (it is Zephyr only)

Reply:
Moving the Kconfig overrides from `nrf/Kconfig.nrf` to `nrfxlib/nrf_security/Kconfig.psa` could be an option though.
The `!TFM_MINIMAL` seems to work better with that setup. The mbedtels_minimal.defconfig file then is able to default it to n.